### PR TITLE
Add Redshift Auth 

### PR DESCRIPTION
**Pull request by hunterwerlla**
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.concurrent.CompletionStage
+
+data class RedshiftSettings(
+    val clusterId: String,
+    val username: String,
+    val credentials: ToolkitCredentialsProvider,
+    val region: AwsRegion
+)
+
+// This is marked as internal but is what we were told to use
```
[the remark this answers is not in the document]
Comment (hunterwerlla) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt:36]: That's much better wording, reworded the other one that way as well.

File: jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuthWidget.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import software.aws.toolkits.jetbrains.services.redshift.extractClusterIdFromUrl
+import software.aws.toolkits.jetbrains.services.redshift.extractRegionFromUrl
+import software.aws.toolkits.jetbrains.ui.AwsAuthWidget
+import software.aws.toolkits.resources.message
+import javax.swing.JPanel
+
+const val CLUSTER_ID_PROPERTY = "AWS.RedshiftClusterId"
+
+class AwsAuthWidget : AwsAuthWidget() {
+    private val clusterIdSelector = JBTextField()
```
[the remark this answers is not in the document]
Comment (hunterwerlla) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuthWidget.kt:22]: Yeah, I want Austin's PR to merge to master before I do that

File: jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        return future {
+            val project = connection.runConfiguration.project
+            val auth = validateConnection(connection)
+            val client = project.awsClient<RedshiftClient>(auth.credentials, auth.region)
+            val credentials = getCredentials(auth, client)
+            DatabaseCredentialsAuthProvider.applyCredentials(connection, credentials, true)
+        }
+    }
+
+    internal fun validateConnection(connection: ProtoConnection): RedshiftSettings {
```
[the remark this answers is not in the document]
Comment (hunterwerlla) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt:54]: These are exposed for testing. We could just test intercept, but we can't inject redshiftclient in a nice way and have to deal with futures

File: jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import software.aws.toolkits.jetbrains.ui.CREDENTIAL_ID_PROPERTY
+import software.aws.toolkits.jetbrains.ui.REGION_ID_PROPERTY
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
+import software.aws.toolkits.resources.message
+import java.util.concurrent.CompletionStage
+
+data class RedshiftSettings(
+    val clusterId: String,
+    val username: String,
+    val credentials: ToolkitCredentialsProvider,
```
[the remark this answers is not in the document]
Comment (hunterwerlla) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt:32]: Sure, changed

File: jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (kiiadi) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt:52]: We could probably push all the logic into a new `internal` class and just have the `AwsAuth: DatabaseAuthProvider` as a wrapper that gives us the lifecycle stuff, the wiring and deals with the futures.

Basically :

```kt
class AwsAuth : DatabaseAuthProvider {
   val delegate = DatabaseCredentialProvider // or whatever

  override fun intercept(...): CompletionStage<ProtoConnection>? {
    return future {
       val credentials = delegate.getCredentials(connection)
    }
  }
}

internal class DatabaseCredentialProvider {
 //move the internal classes here and test them
}

[the remark this answers is not in the document]
Comment (kiiadi) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt:52]: Again - I'm trying to move us to a pattern where "our" logic can be fully tested - and the untested logic is just thin wrappers around the jetbrains entry-points.

[the remark this answers is not in the document]
Comment (hunterwerlla) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuth.kt:52]: Added a test that tests the whole thing, I still like the smaller focused tests as well but we could refactor it for all the tests to be functional

File: jetbrains-ultimate/tst-201/software/aws/toolkits/jetbrains/services/redshift/auth/RedshiftAwsAuthTest.kt [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+    private val redshiftSettings = RedshiftSettings(
+        clusterId = clusterId,
+        username = username,
+        connectionSettings = ConnectionSettings(mock(), region)
+    )
+
+    @Before
+    fun setUp() {
+        MockCredentialsManager.getInstance().addCredentials(credentialId, mockCreds)
```
Comment (abrooksv) [inline jetbrains-ultimate/tst-201/software/aws/toolkits/jetbrains/services/redshift/auth/RedshiftAwsAuthTest.kt:64]: If you dont care about the value of the creds, you can commit the parameter

File: jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuthWidget.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (hunterwerlla) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuthWidget.kt:21]: (Austin's UI test PR https://github.com/aws/aws-toolkit-jetbrains/pull/1711)

File: jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuthWidget.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import com.intellij.ui.components.JBLabel
+import com.intellij.ui.components.JBTextField
+import software.aws.toolkits.jetbrains.services.redshift.RedshiftUtils
+import software.aws.toolkits.jetbrains.ui.AwsAuthWidget
+import software.aws.toolkits.resources.message
+import javax.swing.JPanel
+
+const val CLUSTER_ID_PROPERTY = "AWS.RedshiftClusterId"
+
+class AwsAuthWidget : AwsAuthWidget() {
```
Comment (rli) [inline jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuthWidget.kt:20]: can we rename one of these?

Comment (hunterwerlla) [reply jetbrains-ultimate/src-201/software/aws/toolkits/jetbrains/services/redshift/auth/AwsAuthWidget.kt]: Yep, renamed in a different branch but forgot to pull it in, renamed

File: jetbrains-ultimate/tst-201/software/aws/toolkits/jetbrains/services/redshift/auth/RedshiftAwsAuthTest.kt (hunk 5 above)
Comment (hunterwerlla) [reply jetbrains-ultimate/tst-201/software/aws/toolkits/jetbrains/services/redshift/auth/RedshiftAwsAuthTest.kt:64]: It's easier to use it IMO. We don't validate it but we do have to return the ID so I think it makes more sense to make one than to reach into the default mock credential.

**Pull request merged**